src/rcc.rs

```diff
@@ -1,26 +1,341 @@
 //! Reset and Clock Control
 
-use stm32l4x6::{rcc, RCC};
+// TODO right now the various configure functions reach into rcc directly. This is bad. Add an
+// opaque CR member to RCC, and add methods to CR and BDCR. They should probably take clock source
+// variant arguments.
 
-use ::cmp;
+use stm32l4x6::{rcc, RCC, PWR};
 
 use common::Constrain;
 use time::Hertz;
 use flash::ACR;
 
 impl Constrain<Rcc> for RCC {
+    /// Create an RCC peripheral handle.
+    ///
+    /// Per 6.2, "The MSI is used as system clock source after startup from Reset, configured
+    /// at 4 MHz."
     fn constrain(self) -> Rcc {
+        // Enable write access to the BDCR
+        unsafe { (*PWR::ptr()).cr1.modify(|_,w| w.dbp().set_bit()); }
+        // Write access is (similarly) disabled in CFGR::freeze()
+        // TODO add PWR to the hal to avoid the above nastiness
         Rcc {
             ahb: AHB(()),
             apb1: APB1(()),
             apb2: APB2(()),
             bdcr: BDCR(()),
             csr: CSR(()),
-            cfgr: CFGR { hclk: None, pclk1: None, pclk2: None, sysclk: None }
+            cfgr: CFGR {
+                hclk: None,
+                pclk1: None,
+                pclk2: None,
+                sysclk: clocking::SysClkSource::MSI(
+                    clocking::MediumSpeedInternalRC::new(4_000_000, false)
+                    ),
+            }
         }
     }
 }
 
+pub mod clocking {
+    //! The `clocking` module contains representations of the various objects in the STM32L4x6
+    //! clock tree (see Reference Manual Figs 15 and 16) useful for wiring them up.
+    //!
+    //! There are two main concepts: sources (enum variants) and clocks (structs). For example, the
+    //! SYSCLK clock can be driven by any one of four sources: HSI, MSI, HSI, and PLLCLK. This
+    //! knowledge is encoded in the SysClkSources enum.
+    //!
+    //! Each enum variant contains information about the clock it represents. Some clocks are
+    //! configurable, and thus have fields, and some are not. For example, the LSI
+    //! (LowSpeedInternalRC) clock is always 32 kHz, but the MSI (MediumSpeedInternalRC) clock can
+    //! be configured and thus has a frequency component.
+    //!
+    //! To use them, compose them and feed them to, e.g., sysclk.
+    //!
+    //! ```rust
+    //! let mut rcc = RCC.constrain();
+    //! let msi_clk = clocking::MediumSpeedInternalRC::new(8_000_000, false);
+    //! let sys_clk_src = clocking::SysClkSource::MSI(msi_clk);
+    //! let cfgr = rcc.cfgr.sysclk(sys_clk_src);
+    //! ```
+    //!
+    //! The PLL is a bit more complex because it _is_ a source (`PLLClkOutput`) and also _requires_
+    //! a source (`PLLClkSource`), but you compose the types similarly.
+
+    use super::rcc;
+
+    pub trait InputClock {
+        fn freq(&self) -> u32;
+    }
+
+    /// High-speed internal 16 MHz RC
+    #[derive(Clone, Copy)]
+    pub struct HighSpeedInternal16RC {
```

Comment:
I think maybe it would be better to call it just Clock instead of RC?

Reply:
I tried to make the names match the reference manual as much as possible